### DISCUSSION TITLE
Need help understanding why actions don't [always] trigger

### DISCUSSION THREAD

**Discussion by 0x8000-0000** [taocpp/PEGTL#274] [Q&A] [answered]
Considering this test case:

    #include <tao/pegtl.hpp>
    #include <tao/pegtl/ascii.hpp>
    #include <tao/pegtl/rules.hpp>

    #include <tao/pegtl/contrib/analyze.hpp>
    #include <tao/pegtl/contrib/trace.hpp>

    #include <cassert>
    #include <charconv>
    #include <iostream>

    struct Specification
    {
    };

    namespace
    {
    namespace pegtl = tao::pegtl;

    struct Sep : pegtl::plus<pegtl::ascii::space>
    {
    };

    struct Semi : pegtl::one<';'>
    {
    };

    // Template to generate rule
    // tao::pegtl::seq<Rule0, Separator, Rule1, Separator, Rule2, ... , Separator, RuleN>
    template<typename Separator, typename... Rules>
    struct Interleaved;

    template<typename Separator, typename Rule0, typename... RulesRest>
    struct Interleaved<Separator, Rule0, RulesRest...> : pegtl::seq<Rule0, Separator, Interleaved<Separator, RulesRest...>>
    {
    };

    template<typename Separator, typename Rule0>
    struct Interleaved<Separator, Rule0> : Rule0
    {
    };

    // Note: interleaved<Separator /*, no Rule! */> intentionally not defined.

    template<typename... Rules>
    using SpacedSeq = Interleaved<Sep, Rules...>;

    struct UnsignedInteger : pegtl::plus<pegtl::digit>
    {
    };

    struct KeywordCount : pegtl::keyword<'c', 'o', 'u', 'n', 't'>
    {
    };

    struct KeywordAmount : pegtl::keyword<'a', 'm', 'o', 'u', 'n', 't'>
    {
    };

    struct Ingredient :
        SpacedSeq<pegtl::identifier,
                  pegtl::sor<SpacedSeq<KeywordCount, UnsignedInteger>,
                             SpacedSeq<KeywordAmount, UnsignedInteger, pegtl::identifier>>>
    {
    };

    template<typename Rule>
    struct Statement : pegtl::seq<Rule, pegtl::star<pegtl::ascii::space>, Semi>
    {
    };

    struct Recipe : pegtl::seq<pegtl::list<Statement<Ingredient>, Sep>, pegtl::eof>
    {
    };

    template<typename Rule>
    struct Action : pegtl::nothing<Rule>
    {
    };

    template<>
    struct Action<UnsignedInteger>
    {
        template<typename Input>
        static void apply(const Input& in, Specification& /* specification */)
        {
            std::cout << "Raw matched integer: " << in.string_view() << '\n';

            int quantity = -1;

            [[maybe_unused]] const auto [ptr, ec] = std::from_chars(in.begin(), in.end(), quantity, 10);
            assert(ptr == in.end());

            std::cout << "Parsed integer: " << quantity << '\n';
        }
    };

    template<>
    struct Action<pegtl::identifier>
    {
        template<typename Input>
        static void apply(const Input& in, Specification& /* specification */)
        {
            std::cout << "Found identifier:" << in.string_view() << "\n";
        }
    };

    template<>
    struct Action<Ingredient>
    {
        template<typename Input>
        static void apply(const Input& in, Specification& /* specification */)
        {
            std::cout << "Found ingredient.\n";
        }
    };

    template<>
    struct Action<Recipe>
    {
        template<typename Input>
        static void apply(const Input& in, Specification& /* specification */)
        {
            std::cout << "Found recipe.\n";
        }
    };

    }  // namespace

    static constexpr std::string_view input0 = R"txt(eggs count 2;)txt";

    static constexpr std::string_view input1 = R"txt(eggs count 2;
    flour amount 5 oz;)txt";

    int main()
    {
        const std::size_t issues = tao::pegtl::analyze<Recipe>();
        std::cout << "Analized the grammar and found " << issues << " issues.\n";

        Specification specification;

        {
            std::cout << "Tracing the grammar rules\n";
            pegtl::memory_input in(input1.data(), input1.size(), "");
            pegtl::standard_trace<Recipe>(in);
        }

        {
            std::cout << "Parsing in ?action\n";
            pegtl::memory_input in(input1.data(), input1.size(), "");
            pegtl::parse<Recipe, Action>(in, specification);
        }

        return 0;
    }

When I run it, I get that the input is parsed correctly, but the output is this:

    Parsing in ?action
    Found identifier:eggs
    Found ingredient.
    Found identifier:flour
    Raw matched integer: 5
    Parsed integer: 5
    Found ingredient.
    Found recipe.

So I am missing the unsigned integer "2" on the first line, and the identifier "oz" on the second line.

I have tried creating the parse tree, and the nodes are missing as well.

What am I doing wrong? Thank you!

PS: I am using PEGTL 3.2.1, compiled with Clang13 on Linux, in both C++17 and C++20 modes.

**Comment by d-frey** [comment, marked as answer]
This is due to the `Interleaved`-helper not actually generating the type you think it does. You comment suggests it generates

```c++
pegtl::seq< rule0, sep, rule1, sep, rule2, sep, ... >
```

whereas in fact, it generates

```c++
pegtl::seq< rule0, sep, interleaved< rule1, rule2, ... > >
```

and now your specializations for the actions don't match anymore.

You need a different approach, something like:

```c++
   // Template to generate rule                                                                                                                     
   // tao::pegtl::seq<Rule0, Separator, Rule1, Separator, Rule2, ... , Separator, RuleN>                                                            
   template< typename Result, typename Separator, typename... Rules >
   struct InterleavedImpl;

   template< typename... Results, typename Separator, typename Rule0, typename... RulesRest >
   struct InterleavedImpl< pegtl::seq< Results... >, Separator, Rule0, RulesRest... >
      : InterleavedImpl< pegtl::seq< Results..., Rule0, Sep >, Separator, RulesRest... >
   {};

   template< typename... Results, typename Separator, typename Rule0 >
   struct InterleavedImpl< pegtl::seq< Results... >, Separator, Rule0 >
   {
      using type = pegtl::seq< Results..., Rule0 >;
   };

   template< typename Separator, typename... Rules >
   using Interleaved = typename InterleavedImpl< pegtl::seq<>, Separator, Rules... >::type;
```

**Reply by 0x8000-0000** [reply]
Thank you @d-frey !

Would you consider adding an equivalent functionality as part of the standard PEGTL library? It seems to be quite useful when one wants to parse a seq<> but ignore arbitrary amounts of whitespace. Or is there some other standard mechanism to achieve this?

**Reply by d-frey** [reply]
Maybe. The problem is that it seems generic enough to be useful, but it is different from normal rules. If you consider the current rules, you can always add actions for them. When you have a generator like the above, the rule that is actually used is different. Consider this:

```c++
struct my_rule : seq< rule0, X< rule1, rule2 > > {};
```

Now if `X` is a normal rule, you can specialize an action for

```c++
template<> struct my_action< X< rule1, rule2 > > { ... };
```

but when `X` is a generator, that doesn't work anymore. It breaks the uniformity of our interface.

I *might* add something to the contrib section, let me discuss it with @ColinH when we have more time. (We are currently quite busy)

**Reply by 0x8000-0000** [reply]
Thank you! I am good for now, but I'm thinking other people will search the interwebs for a solution and find the bogus one (which I pasted into my grammar) then try for two days to get it to work (by looking in all the documentation and examples) and debug it, until they figure out to ask or look here.

**Reply by d-frey** [reply]
I added a comment to https://stackoverflow.com/questions/53427551/pegtl-how-to-skip-spaces-for-the-entire-grammar

**Reply by d-frey** [reply]
Actually, there might be a way to make is consistent... I'll still put it to contrib for now, it can still be promoted later if needed.

**Reply by 0x8000-0000** [reply]
@d-frey , you rock!

**Reply by d-frey** [reply]
Done, see my latest commit https://github.com/taocpp/PEGTL/commit/df33fc87cef84e3ab2cff1baf3b7077231d28618

(yeah, I needed one more commit, lets see if I missed anything else)

**Reply by d-frey** [reply]
I was a little bit unhappy about the syntax, so I committed an experiment. What do you think of the new syntax:

```c++
sep< S >::seq< A, B, C, D >
```

which is parsed as

```c++
seq< A, S, B, S, C, S, D >
```

Is this more readable/understandable?

**Reply by 0x8000-0000** [reply]
I looks a bit better, but... is it different from list<R, S> in a way that is surprising. I think the two are somewhat related, and deserve similar enough syntax.

Is this pattern "foo<A>::bar<X, Y>" used anywhere else in PEGTL?

**Reply by d-frey** [reply]
No, which is why it's in contrib and I called it an experiment :)

You are right in that it is unique and other rules do similar stuff without creative syntax.

Going back to `something< S, A, B, C, D >`, what can we choose for `something`? `interleaved` and `separated` (or even `sep`) don't seem very catchy... suggestions, anyone?

**Reply by 0x8000-0000** [reply]
"spaced_tuple" ?

**Reply by d-frey** [reply]
Hm, good start. But `S` is not necessarily a spacer, it's some separator. And the "tuple" is a sequence of rules. OTOH `separated_seq` is a bit long. Why is naming always the hardest part? 🤔

**Reply by 0x8000-0000** [reply]
A "[spacer](https://www.google.com/imgres?imgurl=https%3A%2F%2Fr2-bike.com%2Fmedia%2Fimage%2Fproduct%2F131951%2Flg%2Fpro-spacer-set-ud-carbon-1-1-8_1.jpg&imgrefurl=https%3A%2F%2Fr2-bike.com%2FPRO-Spacer-Set-UD-Carbon-1-1-8_1&tbnid=iqU7p3hLThqGUM&vet=12ahUKEwiWo4jTkrvzAhVQ61MKHcxqDMoQMygxegUIARDPAg..i&docid=4udnW1EgOx7ugM&w=1000&h=750&q=spacer&client=firefox-b-1-d&ved=2ahUKEwiWo4jTkrvzAhVQ61MKHcxqDMoQMygxegUIARDPAg)" is not necessarily made out of invisible stuff. Also [this](https://en.wikipedia.org/wiki/Spacers_and_standoffs)...

**Reply by d-frey** [reply]
I get what you mean, but I think the physical world is different from CS. I don't think that anyone in our domain would call "comma" a spacer, it's a separator.

**Reply by 0x8000-0000** [reply]
How about

    sep< S, Q >

but requiring Q to a sequence?

    seq< A, B, C, D >

Maybe using concepts?

So it could be written inline as:

    sep< S, seq< A, B, C, D > >

**Reply by d-frey** [reply]
No, that breaks the uniformity of our interface. If `seq< A, B, C, D >` is visible in the user's grammar, it must be possible to apply actions to that exact rule.

**Comment by 0x8000-0000** [comment]
Hi @d-frey  - what is a timeline for the next [point] release? I'd like to make use of this new contrib class, but we're trying to stick to official releases. Is this worthy enough of a 3.2.2, and is HEAD stable enough?

Thank you,
florin

**Comment by d-frey** [comment]
We will do a final point release before migrating the license, but we need to clean up some details and adjust the documentation. Due to lack of time, it looks like we will need about 2-3 weeks.

**Reply by 0x8000-0000** [reply]
Understood, thank you.
